### PR TITLE
Set prompt correctly for zsh

### LINKED ISSUE #711: pxssh login to zsh: could not set shell prompt
## Descrpition

When the remote system (user) is using `zsh` as their login shell,
`pexpect.pxssh.ExceptionPxssh: Could not establish connection to host` is
raised from `pxssh.login()`. For example:

```python
 % python3
Python 3.6.9 (default, Jan 26 2021, 15:33:00)
[GCC 8.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import sys
>>> import pexpect.pxssh
>>> p = pexpect.pxssh.pxssh(encoding='utf-8')
>>> p.logfile = sys.stderr
>>> p.login('redacted.example.com', 'root')
Last login: Tue Feb  8 04:17:48 2022 from <redacted>
Copyright (c) 1992-2018 The FreeBSD Project.
Copyright (c) 1979, 1980, 1983, 1986, 1988, 1989, 1991, 1992, 1993, 1994
        The Regents of the University of California. All rights reserved.

<other motd etc>
redacted-hostname#

redacted-hostname#

redacted-hostname#

redacted-hostname#

redacted-hostname# unset PROMPT_COMMAND
PS1='[PEXPECT]\$ '
unset PROMPT_COMMAND
redacted-hostname# PS1='[PEXPECT]\$ '
[PEXPECT]\$ set prompt='[PEXPECT]\$ '
set prompt='[PEXPECT]\$ '
[PEXPECT]\$ Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/<$HOME>/.local/lib/python3.6/site-packages/pexpect/pxssh.py", line 472, in login
    self.before, self.PROMPT,))
pexpect.pxssh.ExceptionPxssh: could not set shell prompt (received: "\x1b[27m\x1b[1m\x1b[m                                                                               \r \r\r\x1b[m\x1b[27m\x1b[24m\x1b[Jredacted-hostname# \x1b[Ku\x08unset PROMPT_COMMAND\r\r\n\x1b[1m\x1b[7m#\x1b[27m\x1b[1m\x1b[m                                                                               \r \r\r\x1b[m\x1b[27m\x1b[24m\x1b[Jredacted-hostname# \x1b[KP\x08PS1='[PEXPECT]\\$ '\r\r\n\x1b[1m\x1b[7m#\x1b[27m\x1b[1m\x1b[m                                                                               \r \r\r\x1b[m\x1b[27m\x1b[24m\x1b[J[PEXPECT]\\$ \x1b[Ks\x08set prompt='[PEXPECT]\\$ '\r\r\n\x1b[1m\x1b[7m#\x1b[27m\x1b[1m\x1b[m                                                                               \r \r\r\x1b[m\x1b[27m\x1b[24m\x1b[J[PEXPECT]\\$ \x1b[K", expected: '\\[PEXPECT\\][\\$\\#] ').
>>> pexpect.__version__
'4.8.0'
>>>
```

## Analysis

From the logged output, it appears that the prompt is actually set by the
command `PS1='[PEXPECT]\$ '`, but it includes the backslash, which is not
expected by `pxssh.PROMPT`. `zsh` doesn't use `\` as an escape character in
prompt strings. The equivalent sequence would be `%(!.#.$)`
([docs](https://zsh.sourceforge.io/Doc/Release/Prompt-Expansion.html#Shell-state)).

## Workaround

If you know the remote is using `zsh`, you can patch `pxssh.PROMPT_SET_SH`:
```python
>>> p = pexpect.pxssh.pxssh(encoding='utf-8')
>>> p.logfile = sys.stderr
>>> p.PROMPT_SET_SH = r"PS1='[PEXPECT]%(!.#.$) '"
>>> p.login('redacted.example.com', 'root')
Last login: Tue Feb  8 04:17:48 2022 from <redacted>
<etc>
redacted-hostname# unset PROMPT_COMMAND
PS1='[PEXPECT]%(!.#.$) '
unset PROMPT_COMMAND
redacted-hostname# PS1='[PEXPECT]%(!.#.$) '
[PEXPECT]# True
>>>
```
Of course having done so you cannot now log in to a non-`zsh` shell.

### PR DESCRIPTION
**Pull request by kbriggs**
Fixes #711

### CODE REVIEW

File: pexpect/pxssh.py [hunk 1]
```diff
@@ -143,6 +143,7 @@ def __init__ (self, timeout=30, maxread=2000, searchwindowsize=None,
         # used to set shell command-line prompt to UNIQUE_PROMPT.
         self.PROMPT_SET_SH = r"PS1='[PEXPECT]\$ '"
         self.PROMPT_SET_CSH = r"set prompt='[PEXPECT]\$ '"
+        self.PROMPT_SET_ZSH = r"PS1='[PEXPECT]%(!.#.$) '"
```
Comment (kbriggs) [inline pexpect/pxssh.py:146]: This is sufficient for regular `zsh` but fails for users with prompt themes from `oh-my-zsh`, which overrides the `PS1` setting.

File: tests/test_pxssh.py [hunk 2]
```diff
@@ -274,5 +274,29 @@ def test_failed_custom_ssh_cmd(self):
...
+        assert ssh.prompt(timeout=10)
+        ssh.logout()
+
+    def test_login_tcsh(self):
+        ssh = pxssh.pxssh()
+        result = ssh.login('server tcsh', 'me', password='s3cret')
+        ssh.sendline('ping')
+        ssh.expect('pong', timeout=10)
+        assert ssh.prompt(timeout=10)
+        ssh.logout()
```
Comment (kbriggs) [inline tests/test_pxssh.py:299]: I wish `@pytest.mark.parametrize` worked in a `unittest`-style test class :(

File: pexpect/pxssh.py [hunk 3]
```diff
@@ -527,11 +528,14 @@ def set_unique_prompt(self):
         self.sendline("unset PROMPT_COMMAND")
         self.sendline(self.PROMPT_SET_SH) # sh-style
         i = self.expect ([TIMEOUT, self.PROMPT], timeout=10)
-        if i == 0: # csh-style
-            self.sendline(self.PROMPT_SET_CSH)
+        if i == 0: # zsh-style
+            self.sendline(self.PROMPT_SET_ZSH)
```
Comment (kbriggs) [inline pexpect/pxssh.py:532]: I'm somewhat agnostic about the order of trying `zsh` vs `csh` here; with MacOS moving to `zsh` as the default it seems like that's likely to be more common these days.

File: pexpect/pxssh.py (hunk 1 above)
Comment (kbriggs) [reply pexpect/pxssh.py]: Fixed in 686fa6f with `prompt restore`.

File: pexpect/pxssh.py (hunk 3 above)
Comment (Red-M) [reply pexpect/pxssh.py:532]: I'd prefer if this stayed in the original order (add zsh after csh) for compatibility reasons for other platforms and to not disturb an already established order.

Comment (kbriggs) [reply pexpect/pxssh.py]: Done

File: tests/test_pxssh.py (hunk 2 above)
Comment (Red-M) [reply tests/test_pxssh.py:299]: It does but not in the way you'd think.
https://docs.python.org/3/library/unittest.html#distinguishing-test-iterations-using-subtests
https://github.com/Red-M/RedSSH/blob/a7d2cf12ebb5f9a4955509c55264212aff14bf2f/tests/test_auth.py#L16

**Comment by Red-M** [comment]
I'll take a full run down of this change on the weekend as I want to get the automated test suite working again for GitHub issues.

**Comment by Red-M** [comment]
I had to close and re-open to get the tests to run. Sorry!

**Comment by Red-M** [comment]
Please fix the issues presented in the tests.
Coverage changes under 1% are fine, do not worry about it.

**Comment by kbriggs** [comment]
> Please fix the issues presented in the tests.

For some reason the merge commit eae7087 merging `master` into the PR removed the execute bit from the mode of `tests/fakessh/ssh`, despite both parent commits still having the execute bit set:
```
% git diff --summary e7cf54a zsh-login-prompt
 create mode 100644 .github/workflows/ci.yml
 mode change 100755 => 100644 tests/fakessh/ssh
% git diff --summary afb85a7 zsh-login-prompt
 mode change 100755 => 100644 tests/fakessh/ssh
```
This prevents the test from executing the mock `ssh` command, so all tests that use it fail.

File: tests/test_pxssh.py (hunk 2 above)
Comment (kbriggs) [reply tests/test_pxssh.py:299]: True, though it doesn't seem to then have the nice don't-stop-at-first-failure and list-parameter-with-testcase-name behaviour. Probably not worth it with only three shells being tested, but if more are added later might be worth considering.

**Comment by kbriggs** [comment]
All `pxssh` tests are now passing. Remaining test failures in 3.11 and pypy3.9 match existing failures in [`master`](https://github.com/pexpect/pexpect/actions/runs/4175699533).

**Pull request merged**